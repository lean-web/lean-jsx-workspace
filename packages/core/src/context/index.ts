interface ComponentContext {
    getId(): string;

    toSource(): string;

    decorate(vnode: SXL.Element): SXL.Element;
}
export type Context<Ctx extends Record<string, unknown>> = Ctx &
    ComponentContext;

export interface SXLGlobalContext {}
