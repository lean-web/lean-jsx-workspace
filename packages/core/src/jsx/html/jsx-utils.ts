export type TextNode = string;
export interface StaticNode extends Omit<SXL.StaticElement, "type"> {
  type: string;
}

export interface FunctionNode extends Omit<SXL.StaticElement, "type"> {
  type: SXL.NodeFactory;
}
export interface ClassNode extends Omit<SXL.StaticElement, "type"> {
  type: SXL.ClassFactory;
}

export function isClass(
  func: string | SXL.NodeFactory | SXL.ClassFactory
): func is SXL.ClassFactory {
  return (
    typeof func === "function" &&
    /^class\s/.test(Function.prototype.toString.call(func))
  );
}

export function isTextNode(jsx: SXL.StaticElement | string): jsx is TextNode {
  return typeof jsx === "string";
}

export function isFunctionNode(
  jsx: SXL.StaticElement | SXL.Children
): jsx is FunctionNode {
  if (typeof jsx === "string" || Array.isArray(jsx)) {
    return false;
  }
  return typeof jsx.type === "function" && !isClass(jsx.type);
}

export function isPromise(
  jsx: SXL.StaticElement | Promise<SXL.StaticElement> | undefined | string
): jsx is Promise<SXL.StaticElement> {
  return typeof jsx !== "string" && !!jsx && "then" in jsx;
}

export function isClassNode(
  jsx: SXL.StaticElement | SXL.Children
): jsx is ClassNode {
  if (typeof jsx === "string" || Array.isArray(jsx)) {
    return false;
  }
  return typeof jsx.type === "function" && isClass(jsx.type);
}

export function isArrayOfNodes(
  jsx: SXL.StaticElement | SXL.Children
): jsx is Array<SXL.StaticElement | string> {
  return Array.isArray(jsx);
}

export function isStaticNode(
  jsx: SXL.StaticElement | string
): jsx is StaticNode {
  return (
    !isTextNode(jsx) &&
    !isArrayOfNodes(jsx) &&
    !isFunctionNode(jsx) &&
    !isClassNode(jsx)
  );
}

export function isFragmentNode(
  jsx: SXL.StaticElement | undefined
): jsx is FunctionNode {
  return !!jsx && /jsxFragment/i.test(jsx.type.toString());
}
