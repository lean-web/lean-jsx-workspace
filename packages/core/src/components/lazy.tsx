/* eslint-disable @typescript-eslint/no-namespace */
import { SXLGlobalContext } from "@/context";

export function toQueryString(
    url: string,
    globalContext?: SXLGlobalContext
): string {
    if (!globalContext) {
        return url;
    }
    return (
        url +
        "?" +
        Object.entries(globalContext)
            .filter(([key, value]) => !!key && !!value)
            .map(([key, value]) => `${key}=${value}`)
            .join("&")
    );
}

interface ComponentArgs extends SXL.Props {
    loading: SXL.Element;
}

abstract class ClassComponent {
    abstract render(): SXL.StaticElement;
    abstract renderLazy(): SXL.AsyncElement;
}

function isPromise(
    jsx: SXL.StaticElement | SXL.AsyncElement | undefined
): jsx is SXL.AsyncElement {
    return !!jsx && "then" in jsx;
}

export class Lazy extends ClassComponent {
    props: ComponentArgs;

    constructor(props: ComponentArgs) {
        super();
        this.props = props;
    }

    async renderLazy(): SXL.AsyncElement {
        if (!this.props.children) {
            throw new Error(
                "There is no child in Lazy to render. This is probably a mistake. If not, please file a bug."
            );
        }
        const allResolved = await Promise.all(this.props.children);
        return <>{allResolved}</>;
    }

    render(): SXL.StaticElement {
        if (isPromise(this.props.loading)) {
            return (<></>) as SXL.StaticElement;
        }
        return this.props.loading;
    }
}

interface PendingResolve<T> {
    isPending: true;
    isResolved: false;
    isError: false;
    value: null;
}

interface ResolvedPromise<T> {
    isPending: false;
    isResolved: true;
    isError: false;
    value: T;
}

type TrackedPromise<T> = PendingResolve<T> | ResolvedPromise<T>;

export interface DynamicController {
    contentId: string;
    Render: (props: SXL.Props) => SXL.Element;
    Api: (props: SXL.Props) => SXL.AsyncElement;
}

export function GetDynamicComponent<T>(
    contentId: string,
    fetcher: () => Promise<T>,
    render: (data: TrackedPromise<T>) => SXL.Element
): DynamicController {
    return {
        Render: (props: SXL.Props) => (
            <dynamic-component
                data-id={toQueryString(contentId, props.globalContext)}
                className="dyn"
            >
                {render({
                    isPending: true,
                    isError: false,
                    isResolved: false,
                    value: null,
                    ...props,
                })}
            </dynamic-component>
        ),
        Api: async (props: SXL.Props) => {
            const data = await fetcher();
            return render({
                isPending: false,
                isError: false,
                isResolved: true,
                value: data,
                ...props,
            });
        },
        contentId,
    };
}

declare global {
    namespace JSX {
        interface IntrinsicElements {
            "dynamic-component": Partial<HTMLElement>;
        }
    }
}
