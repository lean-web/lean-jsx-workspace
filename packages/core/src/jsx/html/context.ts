/* eslint-disable @typescript-eslint/no-unsafe-call */
import { SXLGlobalContext } from "@/context";
import { isPromise } from "./jsx-utils";
import { UIDGenerator } from "./uuid";

type ExtendedEventHandlers<Ctx> = {
    [k in keyof GlobalEventHandlers]: GlobalEventHandlers[k] extends
        | ((this: GlobalEventHandlers, ev: infer E) => any)
        | null
        ? (this: ExtendedEventHandlers<Ctx>, ev: E, ctx: Ctx) => any
        : never;
};

type ClientHandler<
    T extends keyof ExtendedEventHandlers<Ctx>,
    Ctx
> = ExtendedEventHandlers<Ctx>[T];

export type ContextType = "eager" | "lazy";

interface ComponentContext<Ctx> {
    getId(): string;

    toSource(): string;

    decorate(vnode: SXL.Element): SXL.Element;
}
export type Context<Ctx extends Record<string, unknown>> = Ctx &
    ComponentContext<Ctx>;

type KeyHandlerTyple<Ctx, T extends keyof ExtendedEventHandlers<Ctx>> = [
    T,
    ClientHandler<keyof ExtendedEventHandlers<Ctx>, Ctx>
];

export interface ContextFactoryOptions {
    prefix?: string;
}

export class ContextFactory {
    opts?: ContextFactoryOptions;
    globalContext?: SXLGlobalContext;

    idGen: (type?: "PROD" | "DEV" | undefined) => string;

    constructor(
        opts?: ContextFactoryOptions,
        globalContext?: SXLGlobalContext
    ) {
        this.idGen = UIDGenerator.new();
        this.opts = opts;
        this.globalContext = globalContext;
    }
    newContext<K extends Record<string, undefined>>() {
        const ctx = new ContextImpl<K>(this.idGen());
        const handler1: ProxyHandler<ContextImpl<K>> = {
            get(target, prop, receiver) {
                // console.log({ prop, target });
                if (prop in target) {
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
                    return target[prop];
                }
                if (typeof prop === "string") {
                    return target._data[prop];
                }
            },
            set(target, prop, value) {
                if (typeof prop === "string") {
                    target._data[prop] = value;
                    return true;
                }
                return false;
            },
        };

        return new Proxy<ContextImpl<K>>(
            ctx,
            handler1
        ) as unknown as Context<K>;
    }

    buildPlaceholderId(): string {
        const prefix = this.opts?.prefix;
        if (prefix) {
            return `${prefix}-${this.idGen()}`;
        }
        return `placeholder-${this.idGen()}`;
    }
}

export class ContextImpl<Ctx extends Record<string, unknown>>
    implements ComponentContext<Ctx>
{
    private static idGen = UIDGenerator.new();
    private globalContext?: SXLGlobalContext;
    _id: string;
    _handlers: KeyHandlerTyple<Ctx, keyof ExtendedEventHandlers<Ctx>>[] = [];
    _data: Record<string, unknown> = {};

    constructor(id: string, globalContext?: SXLGlobalContext) {
        this._id = id;
        this.globalContext = globalContext;
    }

    getId() {
        return this._id;
    }

    getGlobalContext() {
        return this.globalContext;
    }

    _decorateUnwrappedActions(vnode: SXL.StaticElement) {
        if (!vnode) {
            return;
        }
        const handlers = Object.entries(vnode.props).filter(
            ([key, v]) => /^on/.test(key) || typeof v === "function"
        );

        handlers.forEach(([key, v]) => {
            vnode.props[key] = "";
            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
            this._handlers.push([
                key as keyof GlobalEventHandlers,
                // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                v?.toString(),
            ]);
        });

        return vnode;
    }

    decorate(vnode: SXL.Element): SXL.Element {
        if (!vnode) {
            return vnode;
        }
        if (isPromise(vnode)) {
            return vnode.then((node) => {
                // if (!vnode) {
                //   return vnode;
                // }
                this._decorateUnwrappedActions(node);
                if (this._handlers.length === 0) {
                    return vnode;
                }
                // const dataset = node.props.dataset ?? {};
                if (node.props.dataset) {
                    node.props.dataset[this._id] = "true";
                }
                // node.props.dataset = dataset;
                // console.log({ asyncCtxBefore: node.ctx });
                node.ctx = this as unknown as Context<
                    Record<string, undefined>
                >;
                // console.log({ asyncCtxAfter: node.ctx });
                return node;
            });
        }
        this._decorateUnwrappedActions(vnode);
        if (this._handlers.length === 0) {
            return vnode;
        }
        const dataset = vnode.props.dataset ?? {};
        dataset[this._id] = "true";
        vnode.props.dataset = dataset;
        vnode.ctx = this as unknown as Context<Record<string, undefined>>;

        return vnode;
    }

    toSource() {
        if (this._handlers.length === 0) {
            return "";
        }
        const handlers = this._handlers
            .map(
                ([ev, h]) =>
                    `document.querySelector('[${
                        this._id
                    }]').addEventListener('${ev.replace(
                        /^on/,
                        ""
                    )}', ${h?.toString()})`
            )
            .join(";\n");
        const fns = Object.entries(this._data)
            // .map((entry) => entry[1])
            .filter(
                (entry): entry is [string, Function] =>
                    typeof entry[1] === "function"
            )
            .map(([key, fn]) => `this.${key} =  ${fn.toString()}`)
            .join("\n");
        const source = `<script>
      (function(){
        ${fns}
        ${handlers}
      }).call(${JSON.stringify(this._data)})
    </script>
  `;
        return source;
    }
}
