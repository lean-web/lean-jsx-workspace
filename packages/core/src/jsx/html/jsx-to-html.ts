/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { isStaticNode } from "./jsx-utils";
import { JSDOM } from "jsdom";

function camelToDashed(str: string): string {
    return str.replace(/([a-zA-Z])(?=[A-Z])/g, "$1-").toLowerCase();
}

function normalizeProKey(key: string) {
    return key.replace("className", "class");
}

function normalizePropValue(value: string | number) {
    return typeof value === "number" ? value : `"${value}"`;
}

function styleDeclarationToString(styles: CSSStyleDeclaration): string {
    let styleString = "";
    const entries = Object.entries(styles);
    for (let i = 0; i < entries.length; i++) {
        const [key, value] = entries[i];
        const dashedProperty = camelToDashed(key);
        styleString += `${dashedProperty}: ${value}; `;
    }
    return styleString.trim();
}

function isCSSDeclaration(
    propKey: string,
    propValue: unknown
): propValue is CSSStyleDeclaration {
    return /style/.test(propKey);
}

function flatten(
    props: SXL.Props | Record<string, unknown>
): [string, string | number][] {
    if (Array.isArray(props)) {
        return props;
    }
    return Object.entries(props).flatMap(([key, value]) => {
        if (!value) {
            return [];
        }
        if (/children/.test(key)) {
            return [];
        }
        if (isCSSDeclaration(key, value)) {
            return [[key, styleDeclarationToString(value)]];
        }
        if (typeof value === "object") {
            return flatten(value as Record<string, unknown>);
        }
        if (typeof value !== "string" && typeof value !== "number") {
            throw new Error(
                `Not implemented: Handing property ${key} with value ${value}`
            );
        }
        return [[key, value]];
    });
}

export type OpenAndCloseTags = [string, string];

function serializeObjectToHTML(obj: SXL.StaticElement): [string, string] {
    const { window } = new JSDOM("");
    const { document } = window;

    const { children, ...rest } = obj.props;

    // Create a new element based on the object's tagName property
    const element = document.createElement(obj.type as any);

    // Set attributes from the object
    if (obj.props) {
        for (const [attr, value] of Object.entries(obj.props)) {
            element.setAttribute(attr, value);
        }
    }

    // Get the outerHTML of the created element
    return element.outerHTML.split("><");
}

export class JSXToHTMLUtils {
    static jsxNodeToHTMLTag(jsx: SXL.StaticElement): OpenAndCloseTags {
        if (!isStaticNode(jsx)) {
            throw new Error(
                "Cannot handle JSX nodes with function or class types." +
                    "Please construct the node before calling this method"
            );
        }

        const args = flatten(jsx.props)
            .filter((el) => {
                if (!el) {
                    throw new Error(
                        `Missing prop, ${JSON.stringify(jsx, null, 2)}`
                    );
                }
                return el;
            })
            .map(([key, value]) => {
                try {
                    return `${normalizeProKey(key)}=${normalizePropValue(
                        value
                    )}`;
                } catch (err) {
                    console.error(err, key, value);
                }
            })
            .join(" ");

        return [`<${jsx.type} ${args}>`, `</${jsx.type}>`];

        // const [open, end] = serializeObjectToHTML(jsx);
        // return [`${open}>`, `<${end}`];
    }
}
