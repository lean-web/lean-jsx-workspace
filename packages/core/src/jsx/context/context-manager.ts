import { isPromise, unwrapFragments } from "../html/jsx-utils";
import { UIDGenerator } from "../html/uuid";
import { TrackablePromise } from "../html/stream/stream-utils/trackable-promise";
import { SXLGlobalContext } from "@sxl/core/src/types/context";

interface SyncJSXWrapper {
    id: ContextID;
    element: SXL.StaticElement;
    placeholder: undefined;
    handlers: Array<HandlerPropAndValue>;
    context: SXL.Context<Record<string, unknown>>;
}

interface AsyncJSXWrapper {
    id: ContextID;
    element: TrackablePromise<SXL.StaticElement, unknown>;
    placeholder: SXL.StaticElement;
    handlers: Array<HandlerPropAndValue>;
    context: SXL.Context<Record<string, unknown>>;
}

export type SXLElementWithContext = SyncJSXWrapper | AsyncJSXWrapper;

export function isAsyncElementWithContext(
    element: SXLElementWithContext
): element is AsyncJSXWrapper {
    return (
        element.element instanceof TrackablePromise ||
        isPromise(element.element)
    );
}

export type ContextID = string;

type HandlerPropAndValue = [string, string];

export class LocalContext implements SXL.Context<Record<string, unknown>> {
    [x: string]: unknown;
}

interface ContextManagerOptions {
    sync: boolean;
}

export class ContextManager<G extends SXLGlobalContext> {
    private options: ContextManagerOptions;
    private uiden = UIDGenerator.new();
    private globalContext: G;

    constructor(globalContext: G, options?: ContextManagerOptions) {
        this.globalContext = globalContext;
        this.options = options ?? { sync: false };
    }

    newContext(): SXL.Context<Record<string, unknown>> {
        return new LocalContext();
    }

    fromStaticElement(element: SXL.StaticElement): SXLElementWithContext {
        const id = this.uiden();
        const localCtx = this.newContext();

        return this.processElement(id, localCtx, element);
    }

    fromFunction(element: SXL.FunctionElement): SXLElementWithContext {
        const id = this.uiden();
        const localCtx = this.newContext();

        const newElement = element.type.bind(localCtx)({
            ...element.props,
            children: element.children,
            globalContext: this.globalContext,
        });

        return this.processElement(id, localCtx, newElement);
    }

    fromClass(element: SXL.ClassElement): SXLElementWithContext {
        const id = this.uiden();

        const localCtx = this.newContext();

        const classNode = new element.type({
            ...element.props,
            children: element.children,
            globalContext: this.globalContext,
        });

        return this.processElement(
            id,
            localCtx,
            classNode.renderLazy(),
            classNode.render()
        );
    }

    private processElement(
        id: ContextID,
        context: SXL.Context<Record<string, unknown>>,
        element: SXL.Element,
        placeholder?: SXL.StaticElement
    ): SXLElementWithContext {
        if (isPromise(element)) {
            return {
                id,
                element: new TrackablePromise(
                    element.then((element) => {
                        // if operating in sync mode,
                        // we don't need to wrap the async element
                        // in a template
                        if (this.options.sync) {
                            return element;
                        }
                        if (!element.props.dataset) {
                            element.props.dataset = {};
                        }
                        return {
                            type: "template",
                            props: {
                                id,
                            },
                            children: unwrapFragments(element),
                        };
                    }),
                    id
                ),
                placeholder: this.processPlaceholder(id, placeholder),
                handlers: [],
                context,
            };
        }

        return {
            id,
            element,
            placeholder: undefined,
            handlers: this.processHandlers(id, element),
            context,
        };
    }

    private processHandlers(
        id: ContextID,
        element: SXL.StaticElement
    ): Array<HandlerPropAndValue> {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/ban-types
        const handlers: Array<[string, unknown]> = Object.entries(
            element.props
        ).filter(([key, v]) => /^on/.test(key) || typeof v === "function");

        const _handlers: Array<HandlerPropAndValue> = [];

        handlers.forEach(([key, v]) => {
            let handlerContent = "";
            if (typeof v === "function") {
                handlerContent = v.toString();
            } else if (typeof v === "string") {
                handlerContent = v;
            }
            element.props[key] = "";
            _handlers.push([key as keyof GlobalEventHandlers, handlerContent]);
        });

        if (_handlers.length) {
            element.props.dataset = element.props.dataset ?? {};
            element.props.dataset["data-action"] = id;
        }

        return _handlers;
    }

    private processPlaceholder(
        id: ContextID,
        placehoder?: SXL.StaticElement
    ): SXL.StaticElement {
        return {
            type: "div",
            props: {
                dataset: {
                    ["data-placeholder"]: id,
                },
            },
            children: placehoder ? unwrapFragments(placehoder) : [],
        };
    }
}
