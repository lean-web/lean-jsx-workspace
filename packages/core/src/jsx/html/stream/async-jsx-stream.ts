import { Readable, ReadableOptions } from "stream";
import { ContextFactory } from "../context";
import { JSXToHTMLUtils } from "../jsx-to-html";
import {
    isTextNode,
    isClassNode,
    isFunctionNode,
    isPromise,
} from "../jsx-utils";
import { AsyncChunk } from "./stream-utils/async-chunk";
import { mergeProps } from "./stream-utils/merge-props";
import { JSXStreamOptions } from "./stream-api";

interface HandleElementOptions {
    maybePlaceholder?: SXL.StaticElement;
    insideAsyncResponse?: boolean;
}

export class AsynJSXStream extends Readable {
    private root: JSX.Element;
    private contextFactory: ContextFactory;
    private asyncQueue: AsyncChunk[] = [];
    private called: boolean = false;
    private jsDisabled = false;
    private head: string[] = [];
    private tail: string[] = [];

    constructor(
        root: SXL.Element,
        contextFactory: ContextFactory,
        head: string[],
        tail: string[],
        options?: ReadableOptions
    ) {
        super({ ...options, encoding: "utf-8" });
        this.contextFactory = contextFactory;
        this.root = root;
        this.head = head;
        this.tail = tail;
    }

    _read(_size: number): void {
        if (!this.called) {
            this.called = true;
            this.head.forEach((chunk) => chunk.length > 0 && this.push(chunk));
            this.handleElement(this.root, {});
            queueMicrotask(() => {
                void this.flushAsyncQueue();
            });
        }
    }

    /**
     * Handle Promise<JSX.Element> elements.
     * @param chunk
     * @param loading
     */
    handleAsyncElement(chunk: SXL.AsyncElement, loading?: SXL.StaticElement) {
        const asyncChunk = new AsyncChunk(
            this.contextFactory.buildPlaceholderId(),
            chunk,
            loading
        );
        this.handlePlaceholder(asyncChunk.placeholder);
        this.asyncQueue.push(asyncChunk);
    }

    handleClassElement(classNode: SXL.ClassElement) {
        const loading = classNode.render();
        this.handleAsyncElement(
            Promise.resolve(classNode.renderLazy()),
            loading
        );
    }

    handlePlaceholder(element: SXL.StaticElement) {
        this.handleElement(element, {});
    }

    handleStaticElement(
        chunk: SXL.StaticElement | string,
        maybePlaceholder?: SXL.StaticElement
    ) {
        if (isTextNode(chunk)) {
            this.push(chunk);
        } else if (isClassNode(chunk)) {
            const classNode = new chunk.type({
                ...chunk.props,
                children: chunk.children,
                globalContext: this.contextFactory.globalContext,
            });
            this.handleClassElement(classNode);
        } else if (isFunctionNode(chunk)) {
            const localCtx = this.contextFactory.newContext();
            const newElement = chunk.type.bind(localCtx)({
                ...chunk.props,
                children: chunk.children,
                globalContext: this.contextFactory.globalContext,
            });

            this.handleElement(localCtx.decorate(newElement), {
                maybePlaceholder,
            });
        } else {
            if (!chunk.ctx) {
                const localCtx = this.contextFactory.newContext();
                chunk = localCtx.decorate(chunk) as SXL.StaticElement;
            }

            if (chunk.type === "fragment") {
                chunk.children.forEach((child) => {
                    if (typeof child !== "string") {
                        child.props = {
                            ...child.props,
                            ...mergeProps(chunk as SXL.StaticElement, child),
                        };
                    }

                    this.handleStaticElement(child);
                });
                return;
            }

            const [open, close] = JSXToHTMLUtils.jsxNodeToHTMLTag(chunk);
            this.push(open);
            for (const child of chunk.children) {
                this.handleStaticElement(child);
            }

            this.push(close);
            if (chunk.ctx) {
                this.push(chunk.ctx.toSource());
            }
            this.scheduleAsyncRender();
        }
    }

    handleElement(
        chunk: SXL.Element,
        { maybePlaceholder }: HandleElementOptions
    ): void {
        if (!chunk) {
            // Do nothing
        } else if (isPromise(chunk)) {
            this.handleAsyncElement(chunk, maybePlaceholder);
        } else {
            this.handleStaticElement(chunk, maybePlaceholder);
        }
    }

    get pendingAsyncChunks() {
        return this.asyncQueue.filter(
            (q) => q.element.isFulfilled() && !q.processed
        );
    }

    handleNextAvailableAsync() {
        let handed = 0;
        this.pendingAsyncChunks.forEach((asyncElement) => {
            if (asyncElement.element.isFulfilled()) {
                handed++;
                this.handleElement(asyncElement.element.value, {
                    insideAsyncResponse: true,
                });

                this.push(asyncElement.jsWiring);
                asyncElement.processed = true;
            }
        });
        return handed;
    }

    scheduleAsyncRender() {
        queueMicrotask(() => {
            this.handleNextAvailableAsync();
        });
    }

    flushAsyncQueue() {
        const interval = setInterval(() => {
            const pending = this.asyncQueue.filter(
                (q) => !q.element.isFulfilled() || !q.processed
            );
            if (pending.length === 0) {
                clearInterval(interval);
                this.tail.forEach(
                    (chunk) => chunk.length > 0 && this.push(chunk)
                );
                this.push(null);
            }
            pending.forEach((asyncElement) => {
                if (asyncElement.element.isFulfilled()) {
                    this.handleElement(asyncElement.element.value, {
                        insideAsyncResponse: true,
                    });

                    this.push(asyncElement.jsWiring);
                    asyncElement.processed = true;
                }
            });
        }, 200);

        // while (pending.length) {
        //   const el = pending.shift();
        //   if (el) {
        //     const element = await el.element.promise;
        //     this.handleElement(element);
        //     this.push(el.jsWiring);
        //   }
        // }
    }
}
