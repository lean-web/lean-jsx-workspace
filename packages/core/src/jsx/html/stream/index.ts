import { Readable, ReadableOptions } from "stream";
import { ContextFactory } from "../context";
import { RenderOptions } from "./stream-api";
import { SynJSXStream } from "./syn-jsx-stream";
import { AsynJSXStream } from "./async-jsx-stream";

export function getHTMLStreamFromJSX(
    root: SXL.Element,
    contextFactory: ContextFactory,
    htmlTemplate: [string, string],
    options?: ReadableOptions & RenderOptions
): Readable {
    const { jsDisabled, ...opts } = options ?? {};
    const [start, end] = htmlTemplate;

    if (jsDisabled) {
        return new SynJSXStream(root, contextFactory, [start], [end], {
            ...opts,
            encoding: "utf8",
        });
    }

    return new AsynJSXStream(root, contextFactory, [start], [end], {
        ...opts,
        encoding: "utf8",
    });
}
