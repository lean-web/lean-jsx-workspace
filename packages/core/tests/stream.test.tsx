import { Lazy } from "@/components/lazy";
import { ContextFactory } from "@/jsx/html/context";
import { getHTMLStreamFromJSX } from "@/jsx/html/stream";
import { describe, expect, test } from "@jest/globals";
import pretty from "pretty";

describe("Stream JSX", () => {
    test("Streams JSX as HTML", (done) => {
        async function CostlyComponent(): Promise<SXL.StaticElement> {
            return new Promise((resolve) => {
                setTimeout(() => {
                    resolve(
                        <div className="costly">
                            I'm costly content
                            <button
                                onclick={() => console.log("Inside costly")}
                            >
                                Click me
                            </button>
                        </div>
                    );
                });
            });
        }

        const jsxElement = (
            <div>
                <h1>Hello world</h1>
                <CostlyComponent />
                <p>After async</p>
            </div>
        );

        let result = ``;
        const jsxToHTMLStream = getHTMLStreamFromJSX(
            jsxElement,
            new ContextFactory(),
            ["<body>", "</body>"]
        );
        // jsxToHTMLStream.write(jsxElement);
        jsxToHTMLStream.on("data", (data) => {
            result += typeof data === "object" ? JSON.stringify(data) : data;
        });
        jsxToHTMLStream.on("end", () => {
            expect(pretty(result)).toMatchSnapshot();
            done();
        });
    }, 2000);

    test("Streams JSX as HTML into another stream", (done) => {
        async function Async() {
            return <p>I'm async</p>;
        }

        const jsxElement = (
            <div>
                <h1>Hello world</h1>
                <Async />
                <p>After async</p>
            </div>
        );

        let result = ``;

        const jsxToHTMLStream = getHTMLStreamFromJSX(
            jsxElement,
            new ContextFactory(),
            ["<body>", "</body>"]
        );
        // jsxToHTMLStream.write(jsxElement);
        jsxToHTMLStream.on("data", (data) => {
            result += typeof data === "object" ? JSON.stringify(data) : data;
        });
        jsxToHTMLStream.on("end", () => {
            expect(pretty(result)).toMatchSnapshot();
            done();
        });
        // jsxToHTMLStream.end();
    });

    test.skip("Streams JSX as HTML with handlers", () => {
        function Button() {
            return <button onclick={(ev) => console.log(ev)}>Click me!</button>;
        }

        const jsxElement = (
            <div>
                <h1>Hello world</h1>
                <Button />
                <p>After Button</p>
            </div>
        );

        let result = ``;

        const jsxToHTMLStream = getHTMLStreamFromJSX(
            jsxElement,
            new ContextFactory(),
            ["<body>", "</body>"]
        );

        jsxToHTMLStream.on("data", (data) => {
            result +=
                typeof data === "object" ? JSON.stringify(data) : data + "\n";
        });
    });

    test("Streams JSX as HTML with nested async functions", (done) => {
        async function AsyncButton() {
            return (
                <button onclick={() => console.log("Inside costly")}>
                    Click me
                </button>
            );
        }
        async function CostlyComponent(): Promise<SXL.StaticElement> {
            return new Promise((resolve) => {
                setTimeout(() => {
                    resolve(
                        <div className="costly">
                            I'm costly content
                            <AsyncButton />
                        </div>
                    );
                });
            });
        }

        const jsxElement = (
            <div>
                <h1>Hello world</h1>
                <Lazy loading={<p>Loading...</p>}>
                    <CostlyComponent />
                </Lazy>
                <p>After async</p>
            </div>
        );

        let result = ``;
        const jsxToHTMLStream = getHTMLStreamFromJSX(
            jsxElement,
            new ContextFactory(),
            ["<body>", "</body>"]
        );
        // jsxToHTMLStream.write(jsxElement);
        jsxToHTMLStream.on("data", (data) => {
            result += typeof data === "object" ? JSON.stringify(data) : data;
        });
        jsxToHTMLStream.on("end", () => {
            expect(pretty(result)).toMatchSnapshot();
            done();
        });
    });

    test("Lazy elements display placeholders", (done) => {
        async function CostlyComponent(): Promise<SXL.StaticElement> {
            return new Promise((resolve) => {
                setTimeout(() => {
                    resolve(<div className="costly">I'm costly content</div>);
                });
            });
        }

        const jsxElement = (
            <Lazy loading={<p>Loading...</p>}>
                <CostlyComponent />
            </Lazy>
        );

        let result = ``;
        const jsxToHTMLStream = getHTMLStreamFromJSX(
            jsxElement,
            new ContextFactory(),
            ["<body>", "</body>"]
        );
        // jsxToHTMLStream.write(jsxElement);
        jsxToHTMLStream.on("data", (data) => {
            result += typeof data === "object" ? JSON.stringify(data) : data;
        });
        jsxToHTMLStream.on("end", () => {
            expect(result.includes("Loading")).toBeTruthy();
            expect(pretty(result)).toMatchSnapshot();
            done();
        });
    });
});
