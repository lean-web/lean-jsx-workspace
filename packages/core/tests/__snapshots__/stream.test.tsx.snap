// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stream JSX Lazy elements display placeholders 1`] = `
"<body>
  <p data-placeholder="placeholder-element-0">Loading...</p><template id="placeholder-element-0">
    <div data-placeholder="placeholder-element-6"></div>
  </template>
  <script>
    sxl.fillPlaceHolder("placeholder-element-0");
  </script>
  <template id="placeholder-element-6">
    <div class="costly">I'm costly content</div>
  </template>
  <script>
    sxl.fillPlaceHolder("placeholder-element-6");
  </script>
</body>"
`;

exports[`Stream JSX Streams JSX as HTML 1`] = `
"<body>
  <div>
    <h1>Hello world</h1>
    <div data-placeholder="placeholder-element-3"></div>
    <p>After async</p>
  </div><template id="placeholder-element-3">
    <div class="costly">I'm costly content<button element-8="true">Click me</button>
      <script>
        (function() {

          document.querySelector('[element-8]').addEventListener('click', () => console.log("Inside costly"))
        }).call({})
      </script>
    </div>
  </template>
  <script>
    sxl.fillPlaceHolder("placeholder-element-3");
  </script>
</body>"
`;

exports[`Stream JSX Streams JSX as HTML into another stream 1`] = `
"<body>
  <div>
    <h1>Hello world</h1>
    <div data-placeholder="placeholder-element-3"></div>
    <p>After async</p>
  </div><template id="placeholder-element-3">
    <p>I'm async</p>
  </template>
  <script>
    sxl.fillPlaceHolder("placeholder-element-3");
  </script>
</body>"
`;

exports[`Stream JSX Streams JSX as HTML with nested async functions 1`] = `
"<body>
  <div>
    <h1>Hello world</h1>
    <p data-placeholder="placeholder-element-2">Loading...</p>
    <p>After async</p>
  </div><template id="placeholder-element-2">
    <div data-placeholder="placeholder-element-9"></div>
  </template>
  <script>
    sxl.fillPlaceHolder("placeholder-element-2");
  </script>
  <template id="placeholder-element-9">
    <div class="costly">I'm costly content<div data-placeholder="placeholder-element-14"></div>
    </div>
  </template>
  <script>
    sxl.fillPlaceHolder("placeholder-element-9");
  </script>
  <template id="placeholder-element-14"><button element-13="true">Click me</button>
    <script>
      (function() {

        document.querySelector('[element-13]').addEventListener('click', () => console.log("Inside costly"))
      }).call({})
    </script>
  </template>
  <script>
    sxl.fillPlaceHolder("placeholder-element-14");
  </script>
</body>"
`;
